File: 2343-count-unguarded-cells-in-the-grid/count-unguarded-cells-in-the-grid.py
class Solution:
    def countUnguarded(self, m: int, n: int, guards: List[List[int]], walls: List[List[int]]) -> int:
        dp = [[0] * n for _ in range(m)]
        
        for x, y in guards:
            dp[x][y] = 3  
        for x, y in walls:
            dp[x][y] = 3

        dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)]

        for gx, gy in guards:
            for dx, dy in dirs:
                x, y = gx, gy

                while True:
                    x += dx
                    y += dy

                    if x < 0 or y < 0 or x >= m or y >= n or dp[x][y] == 3:
                        break

                    dp[x][y] = 1

        ans = 0
        for arr in dp:
            zero = arr.count(0)
            ans += zero

        return ans

File: 2343-count-unguarded-cells-in-the-grid/count-unguarded-cells-in-the-grid.cs
public class Solution {
    public int CountUnguarded(int m, int n, int[][] guards, int[][] walls) {
        HashSet<(int, int)> blockedCells = new HashSet<(int, int)>();
        HashSet<(int, int)> guardedCells = new HashSet<(int, int)>();

        foreach (var wall in walls) {
            blockedCells.Add((wall[0], wall[1]));
        }
        foreach (var guard in guards) {
            blockedCells.Add((guard[0], guard[1]));
        }

        int[] dx = { -1, 0, 1, 0 };
        int[] dy = { 0, 1, 0, -1 };

        foreach (var guard in guards) {
            int x = guard[0];
            int y = guard[1];

            for (int d = 0; d < 4; d++) {
                int nx = x;
                int ny = y;

                while (true) {
                    nx += dx[d];
                    ny += dy[d];

                    if (nx < 0 || ny < 0 || nx >= m || ny >= n || blockedCells.Contains((nx, ny))) {
                        break;
                    }

                    guardedCells.Add((nx, ny));
                }
            }
        }

        int count = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                var cell = (i, j);
                if (!blockedCells.Contains(cell) && !guardedCells.Contains(cell)) {
                    count++;
                }
            }
        }

        return count;
    }
}